### ISSUE TITLE
aws_sqs sink high memory consumption. 

### ISSUE THREAD

**Issue by makasim** [vectordotdev/vector#18884]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

We face a problem with high memory consumption in a vector sidecar. The sidecar memory consumption grows to 600-700Mb and stabilizes at that level. The sidecar is used to collect logs. We have about 1000 RPS incoming traffic on the fluent source, the traffic is constant with no big spikes. 

Some observations:
* There are no discarded events or errors in any components. 
* The buffers stay empty. 
* Metrics cardinality around 70 and not growing. 
* Received traffic equals to send one.
* The problem exists both on v0.31.0 and latest (v0.33.0) vector versions.
* Disabling acknowledgment has no effect. The problem exists.
* We also tried console and amqp sinks. Memory does not grow, stays around 120Mb.
* Memory grows only if there is an incoming load. 
* Memory never goes down, either stays (if no load) or grows to some point. 
* Setting `expire_metrics_secs` option has no effect. The problem still exist.
* Tested with v0.23.3. The issue still exists, but memory grows slower.

### Configuration

```text
[api]
enabled = true

[acknowledgements]
enabled = true

[sources.app_container]
type = "fluent"
address = "0.0.0.0:24224"

[sinks.sqs_logger_applogs]
type = "aws_sqs"
acknowledgements.enabled = false
inputs = ["app_container"]
queue_url = "${QUEUE_URL}"
encoding.codec = "json"
```

### Version

v0.33.0

### Additional Context

Memory grow graph and distribution among sidecars:

<img width="1452" alt="Screenshot 2023-10-19 at 12 42 23" src="https://github.com/vectordotdev/vector/assets/143206/f05caa27-1b02-453a-9332-819550840383">
<img width="1456" alt="Screenshot 2023-10-19 at 12 42 43" src="https://github.com/vectordotdev/vector/assets/143206/2dc8584a-4112-4c61-9ffa-317ba9563e5c">

Memory consumption with amqp config:

<img width="1453" alt="Screenshot 2023-10-19 at 12 57 20" src="https://github.com/vectordotdev/vector/assets/143206/1d37cbd4-47fc-462a-a482-b065d0fa46ca">

Amqp config:

```
[api]
enabled = true

[acknowledgements]
enabled = false

[sources.app_container]
type = "fluent"
address = "0.0.0.0:24224"

[sinks.rmq_logger_applogs]
type = "amqp"
inputs = ["app_container"]
connection_string = "amqp://${RMQ_USER}:${RMQ_PASSWORD}@${RMQ_HOST}:${RMQ_PORT:-5672}/${RMQ_VHOST}?timeout=10"
exchange = "${RMQ_EXCHANGE_APP_LOGS}"
encoding.codec = "json"
```

Console config:

```
[api]
enabled = true

[acknowledgements]
enabled = false

[sources.app_container]
type = "fluent"
address = "0.0.0.0:24224"

[sinks.console]
type = "console"
inputs = [ "logs_enhanced" ]
inputs = ["app_container"]
target = "stdout"
encoding.codec = "json"
```

**Comment by makasim** [comment]
This is an essential issue for me. I'd be happy to fix it myself but I have almost zero knowledge in Rust. 

I'd be happy to see the issue fixed and I am willing to pay someone to fix it. 

@jszwedko @bruceg do you have any idea who could help me?

**Comment by makasim** [comment]
I've run vector (fluent\sqs) with [allocation tracking enabled](https://vector.dev/blog/tracking-allocations/). Memory use does not grow (see screen) 

<img width="1475" alt="Screenshot 2023-10-23 at 16 07 22" src="https://github.com/vectordotdev/vector/assets/143206/ad93882d-65a4-4830-aadb-a1b99843ee21">



But the overall memory consumption grows as before and it may have a correlation with "bytes out" column.  

<img width="957" alt="Screenshot 2023-10-23 at 16 07 15" src="https://github.com/vectordotdev/vector/assets/143206/3ec7e86a-aaab-4609-b2bc-fdf2435f0c1a">

**Comment by makasim** [comment]
Setting buffer max events and dropping the newest has no effect on memory consumption. It still grows. 

```
[sinks.sqs_logger_applogs]
type = "aws_sqs"
buffer.type = "memory"
buffer.when_full = "drop_newest"
buffer.max_events = 1000
inputs = ["app_container"]
queue_url = "https://sqs.us-west-2.amazonaws.com/xxx/test-queue"
encoding.codec = "native_json"

```

**Comment by makasim** [comment]
Can reproduce it on linux/amd64, darwin/amd64

**Comment by makasim** [comment]
A note for anyone considering vector as logging sidecar.

It looks like that vector is not design to be used as a logging sidecar. A loging sidecar must consume as less memory as possible to not interfere with the application. For the sidecar, memory consumption is even more important than high throuput. Vector obviosly cannot limit its memory consumption. 

Also, vector will exit in case a sink lost a connection (got it with amqp sink). Causing the app being killed (not graceful shutdown) and restart too.

I failed to find a feature that would allow me to stop app first, drain logs, and only than stop vector.

**Comment by StephenWakely** [comment]
> Also, vector will exit in case a sink lost a connection (got it with amqp sink). Causing the app being killed (not graceful shutdown) and restart too.

This is a bug that is referenced here #17182 and needs fixing.

**Comment by StephenWakely** [comment]
I'm unable to reproduce this. 

Running SQS with localstack in Docker:

```sh
docker run -e SERVICES=sqs -d -p 4566:4566 --name nork localstack/localstack-full:0.11.6
```

And this conf:

```yml
api:
  enabled: true

sources:
  in:
    type: demo_logs
    format: json
    interval: 0
    decoding:
      codec: bytes
    framing:
      method: bytes

sinks:
  out:
    type: "aws_sqs"
    buffer:
      type: "memory"
      when_full: "drop_newest"
      max_events: 1000
    inputs: ["in"]
    acknowledgements:
      enabled: false
    endpoint: http://localhost:4566/
    queue_url: http://localhost:4566/000000000000/sample-queue
    region: localstack
    encoding:
      codec: "native_json"
```

I get fairly consistent memory usage of 140mb.

With default buffer settings memory usage seems to be around 280mb.

Curious if you are seeing similar with those settings?

**Comment by makasim** [comment]
Unfortunately, 280Mb for log sidecar is not acceptable in my case. The whole instance has 256Mb hard limit. That's enough for my workload and growing the instance to just process logs seems unreasonable. 

So, I decided to experiment and write an in-house solution. The test was to perform 500RPS and generate around 200kb/s traffic. It is close to the production traffic (I posted screens before). 

Here's the stats:
Vector: 500Mb RAM, 0.05 CPU
Fluent: 154Mb RAM, 0.1 CPU
In-house: 20Mb, 0.02 CPU

Closing the issue, because from what I can see such memory consumption is not a bug.
